Change nest to cloud push

awarecan opened the pull request:
## Description:
This is the follow up of python-nest upgrade #14638. After python-nest change to stream API, we can change nest component to Cloud Push mode.

**Related issue (if applicable):** fixes #10442 

**Pull request in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io) with documentation (if applicable):**
home-assistant/home-assistant.github.io#5442

## Example entry for `configuration.yaml` (if applicable):
No configuration changes
```yaml

```

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io)

If the code communicates with devices, web services, or third-party tools:
  - [x] New dependencies have been added to `requirements_all.txt` by running `script/gen_requirements_all.py`.

awarecan committed:
Change nest component to Cloud Push

Change sensors.nest, binary_sensors.nest and climate.nest to push mode
nest camera still need poll to update snapshot image
Also change nest component to async

awarecan committed:
Flake8 lint

awarecan committed:
Fix async_notify_errors, it is not a coroutine

awarecan committed:
Fix pylint

awarecan committed:
Fix pylint, function name should shall shorter than 32

MartinHjelmare commented on homeassistant/components/nest.py:
```diff
@@ -78,44 +103,61 @@ def nest_configuration_callback(data):
-    discovery.load_platform(hass, 'climate', DOMAIN, {}, config)
-    discovery.load_platform(hass, 'camera', DOMAIN, {}, config)
+    _LOGGER.debug("proceeding with discovery -- climate")
+    await discovery.async_load_platform(hass, 'climate', DOMAIN, {}, config)
```
Use `async_add_job` to schedule the discovery to avoid a deadlock where the component waits for the platform to be setup which waits for the component to be setup.

MartinHjelmare replied:
Also use a for loop to that loads all the platforms. Iterate over component and platform config discovery info.

awarecan replied:
Done

MartinHjelmare commented on homeassistant/components/climate/nest.py:
```diff
@@ -97,6 +100,16 @@ def __init__(self, structure, device, temp_unit):
+
+    async def async_nest_update_event_handler(self, event):
+        """Update device state."""
+        await self.async_device_update()
```
```py
self.async_schedule_update_ha_state(True)
```

awarecan replied:
After change to use dispatcher, can I use `awit self.async_update_ha_state(True)` here?
As my understanding, dispatcher will create a job for the callback, if I use `async_schedule_update_ha_state`, it will add another job, seems unnecessary

MartinHjelmare replied:
Yes, that should also work. Then keep this method a coroutine.

awarecan replied:
Done

MartinHjelmare commented on homeassistant/components/climate/nest.py:
```diff
@@ -97,6 +100,16 @@ def __init__(self, structure, device, temp_unit):
+    async def async_nest_update_event_handler(self, event):
+        """Update device state."""
+        await self.async_device_update()
+        await self.async_update_ha_state()
```
Remove this.

MartinHjelmare commented on homeassistant/components/climate/nest.py:
```diff
@@ -97,6 +100,16 @@ def __init__(self, structure, device, temp_unit):
+        """Do not need poll thanks using Nest streaming API."""
+        return False
+
+    async def async_nest_update_event_handler(self, event):
```
This should be a callback and not a coroutine. Decorate it with `@callback` imported from core.py.

MartinHjelmare commented on homeassistant/components/climate/nest.py:
```diff
@@ -37,11 +37,14 @@ def setup_platform(hass, config, add_devices, discovery_info=None):
+    add_devices(all_devices, True)
+
+    for device in all_devices:
+        hass.bus.listen(EVENT_NEST_UPDATE,
```
See above.

MartinHjelmare commented on homeassistant/components/binary_sensor/nest.py:
```diff
@@ -8,9 +8,9 @@
 
 from homeassistant.components.binary_sensor import (BinarySensorDevice)
+from homeassistant.components.nest import DATA_NEST, EVENT_NEST_UPDATE
 from homeassistant.components.sensor.nest import NestSensor
```
It's not allowed to depend on another platform from a platform. You can define a common base nest entity class instead in the nest component and import that into the nest platforms.

awarecan replied:
That is old code. Do you want to me fix that in this PR?

MartinHjelmare replied:
It can be done separately. I mentioned it since the event handler that is defined in the sensor nest base class is now used in this binary sensor platform.

awarecan replied:
Will do that later

MartinHjelmare commented on homeassistant/components/nest.py:
```diff
@@ -51,23 +55,44 @@
+    so specific an executor to save default thread pool.
+    """
+    _LOGGER.debug("listening nest.update_event")
+    with ThreadPoolExecutor(max_workers=1) as executor:
```
I don't think we should do this. Can't python-nest allow a callback to be registered? The callback should be called by python-nest when a state update is needed.

Otherwise waiting for the event should be done by a coroutine asynchronously. But the current event is a `threading.Event` so I don't think that will work as it is now.

awarecan replied:
My code works. `run_in_executor(event.wait)` is the way to glue `threading.Event` and `asyncio`, and `async_nest_update_event_broker` is a coroutine.

MartinHjelmare replied:
Yes, but it's blocking a thread, which isn't good.

awarecan replied:
That is why I am creating a new executor, so that I just block my own thread, won't affect others.

python-nest's author is not responsive like you guys, the stream API changes have to wait months to get review and merged. I think we better to use whatever we have now.

BTW, I really dig myself into a hole. That update_event was introduce by myself to the python-nest, I should allow caller pass in an optional event object

MartinHjelmare replied:
The new executor is what I don't like.

Maybe @balloob has a better idea than me how to do this?

balloob replied:
The dev of python-nest seems active right now, if it's a small change should probably be able to get in quickly?

awarecan replied:
What is the prefer way? An simple callback function enough?

awarecan replied:
@balloob I am rethinking that under the influence of caffeine, I feel like current implement is not bad from python-nest perspective. As a service provider, it doesn't know what its client will put in the callback, it might be better just use thread event to loosen the coulping.

My next  change for python-nest will be allow only monitor part of the data, so I may need multiple update event, and I may try to async-ized whole package, at that time, I will provide an better interface for coroutine call back.

balloob replied:
Ok, that sounds good. If you go async the whole package, consider allowing passing in an aiohttp client session, that way it is best suited for HASS 😉

awarecan committed:
Use dispatcher helper instead event bus

awarecan committed:
Use async_update_ha_state(True)

awarecan committed:
Refactoring load_platform

Move service registration into async_setup_nest(),
 resolve an issue that before the first time configuration done,
 set_mode service should not be registered

awarecan committed:
Fix an issue that authorization failure may leave a blocked thread

awarecan commented on homeassistant/components/binary_sensor/nest.py:
```diff
@@ -104,6 +104,10 @@ def setup_platform(hass, config, add_devices, discovery_info=None):
     add_devices(sensors, True)
 
+    for sensor in sensors:
+        hass.bus.listen(EVENT_NEST_UPDATE,
```
Done

awarecan committed:
Pylinting

MartinHjelmare commented on homeassistant/components/climate/nest.py:
```diff
@@ -37,11 +38,14 @@ def setup_platform(hass, config, add_devices, discovery_info=None):
+    add_devices(all_devices, True)
+
+    for device in all_devices:
+        async_dispatcher_connect(hass, SIGNAL_NEST_UPDATE,
```
This should be done inside the entity coroutine method `async_added_to_hass`. Otherwise there might be an error if the entity state update methods are called before the entity has been added to home assistant.

balloob commented on homeassistant/components/climate/nest.py:
```diff
@@ -97,6 +101,15 @@ def __init__(self, structure, device, temp_unit):
+
+    async def async_nest_update_callback(self):
+        """Update device state."""
+        await self.async_update_ha_state(True)
```
We should never await this. Just call `self.self.async_schedule_update_ha_state(True)` (no need to await)

MartinHjelmare replied:
@balloob It's the coroutine `async_update_ha_state` not the callback `async_schedule_update_ha_state`.

awarecan committed:
async_nest_update_callback => async_update_state to avoid confusion

awarecan committed:
Move signal handler register to async_added_to_hass

awarecan committed:
Better handle nest api error

awarecan committed:
Remove unnecessary register for binary_sensor

awarecan committed:
Remove unused import

balloob commented on homeassistant/components/nest.py:
```diff
@@ -51,23 +55,44 @@
+    _LOGGER.debug("listening nest.update_event")
+    with ThreadPoolExecutor(max_workers=1) as executor:
+        while True:
+            await hass.loop.run_in_executor(executor, nest.update_event.wait)
```
The problem with this is that now Home Assistant is unable to shut down until an event comes in from Nest.

awarecan replied:
Already think about it, next line will check `if hass.is_running`, and at line 172 `shut_down` method, I set the event to allow this statement continue

awarecan commented:
Caught a bug  in python-nest, jkoelker/python-nest#136. We better wait this fix got merged.

This error most likely no to affect current poll implementation, since all non-coroutine update will execute sequentially, the thread re-entry won't be triggered

awarecan committed:
Upgrade to python-nest 4.0.1

Fix a thread race condition issue

awarecan commented:
@balloob @MartinHjelmare upstream package has updated, this PR is good for merge now.

balloob commented on homeassistant/components/climate/nest.py:
```diff
@@ -168,18 +182,24 @@ def is_away_mode_on(self):
+            _LOGGER.error("An error occurred while setting temperature: %s",
+                          api_error)
+            # restore target temperature
+            self.hass.add_job(self.async_update_state)
```
I don't really see why we would have to restore target temperature? Where was it overridden?

Also, just call `hass.schedule_update_ha_state(True)`

balloob commented on homeassistant/components/climate/nest.py:
```diff
@@ -97,6 +97,20 @@ def __init__(self, structure, device, temp_unit):
+        """Do not need poll thanks using Nest streaming API."""
+        return False
+
+    async def async_update_state(self):
```
Can you define this function inside `async_added_to_hass`, I wouldn't want to have it accidentally clash with a future function on entity with a similiar name.

balloob commented on homeassistant/components/climate/nest.py:
```diff
@@ -97,6 +97,20 @@ def __init__(self, structure, device, temp_unit):
+
+    async def async_update_state(self):
+        """Update device state."""
+        await self.async_update_ha_state(True)
```
Please only call `async_schedule_update_ha_state(True)`

balloob committed:
Address my own comments

balloob committed:
Address my own comment

rossdargan commented:
The docs still say this component used cloud polling - Is that still correct?

awarecan commented:
@balloob @MartinHjelmare could you merge the doc PR home-assistant/home-assistant.github.io#5442

MartinHjelmare commented:
Done.
